File: CSharp/Neural.cs
/*
Licensed under the MIT License.
Copyright 2023-2025 Daniel Lidstrom
*/

namespace Neural;

public static class ActivationFunctions
{
    public static double Sigmoid(double f)
    {
        return 1.0 / (1.0 + Math.Exp(-f));
    }

    public static double SigmoidPrim(double f)
    {
        return f * (1.0 - f);
    }
}

public record Network(
    int InputCount,
    int HiddenCount,
    int OutputCount,
    double[] WeightsHidden,
    double[] BiasesHidden,
    double[] WeightsOutput,
    double[] BiasesOutput)
{
    public double[] Predict(double[] input)
    {
        double[] y_hidden = new double[HiddenCount];
        double[] y_output = new double[OutputCount];
        return Predict(input, y_hidden, y_output);
    }

    public double[] Predict(double[] input, double[] y_hidden, double[] y_output)
    {
        for (int c = 0; c < HiddenCount; c++)
        {
            double sum = 0.0;
            for (int r = 0; r < InputCount; r++)
            {
                sum += input[r] * WeightsHidden[r * HiddenCount + c];
            }

            y_hidden[c] = ActivationFunctions.Sigmoid(sum + BiasesHidden[c]);
        }

        for (int c = 0; c < OutputCount; c++)
        {
            double sum = 0.0;
            for (int r = 0; r < HiddenCount; r++)
            {
                sum += y_hidden[r] * WeightsOutput[r * OutputCount + c];
            }

            y_output[c] = ActivationFunctions.Sigmoid(sum + BiasesOutput[c]);
        }

        return y_output;
    }
}

public record Trainer(
    Network Network,
    double[] Hidden,
    double[] Output,
    double[] GradHidden,
    double[] GradOutput)
{
    public static Trainer Create(Network network, int hiddenCount, int outputCount)
    {
        double[] hidden = new double[hiddenCount];
        double[] output = new double[outputCount];
        double[] gradHidden = new double[hiddenCount];
        double[] gradOutput = new double[outputCount];
        return new Trainer(network, hidden, output, gradHidden, gradOutput);
    }

    public static Trainer Create(int inputCount, int hiddenCount, int outputCount, Func<double> rand)
    {
        double[] hidden = new double[hiddenCount];
        double[] output = new double[outputCount];
        double[] gradHidden = new double[hiddenCount];
        double[] gradOutput = new double[outputCount];
        double[] weightsHidden = Array.ConvertAll(new double[inputCount * hiddenCount], v => rand.Invoke() - 0.5);
        double[] biasesHidden = new double[hiddenCount];
        double[] weightsOutput = Array.ConvertAll(new double[hiddenCount * outputCount], v => rand.Invoke() - 0.5);
        double[] biasesOutput = new double[outputCount];
        Network network = new(inputCount, hiddenCount, outputCount, weightsHidden, biasesHidden, weightsOutput, biasesOutput);
        return new Trainer(network, hidden, output, gradHidden, gradOutput);
    }

    public void Train (double[] input, double[] y, double lr)
    {
        _ = Network.Predict(input, Hidden, Output);
        for (int c = 0; c < Network.OutputCount; c++)
        {
            GradOutput[c] = (Output[c] - y[c]) * ActivationFunctions.SigmoidPrim(Output[c]);
        }

        for (int r = 0; r < Network.HiddenCount; r++)
        {
            double sum = 0.0;
            for (int c = 0; c < Network.OutputCount; c++)
            {
                sum += GradOutput[c] * Network.WeightsOutput[r * Network.OutputCount + c];
            }

            GradHidden[r] = sum * ActivationFunctions.SigmoidPrim(Hidden[r]);
        }

        for (int r = 0; r < Network.HiddenCount; r++)
        {
            for (int c = 0; c < Network.OutputCount; c++)
            {
                Network.WeightsOutput[r * Network.OutputCount + c] -= lr * GradOutput[c] * Hidden[r];
            }
        }

        for (int r = 0; r < Network.InputCount; r++)
        {
            for (int c = 0; c < Network.HiddenCount; c++)
            {
                Network.WeightsHidden[r * Network.HiddenCount + c] -= lr * GradHidden[c] * input[r];
            }
        }

        for (int c = 0; c < Network.OutputCount; c++)
        {
            Network.BiasesOutput[c] -= lr * GradOutput[c];
        }

        for (int c = 0; c < Network.HiddenCount; c++)
        {
            Network.BiasesHidden[c] -= lr * GradHidden[c];
        }
    }
}

File: Zig/src/neural.zig
const std = @import("std");

const RuntimeError: type = error{
    InputSizeMismatch,
    HiddenSizeMismatch,
    OutputSizeMismatch,
};

fn sigmoid(x: f64) f64 {
    return 1 / (1 + std.math.exp(-x));
}

fn sigmoidPrime(d: f64) f64 {
    return d * (1 - d);
}

pub const Network = struct {
    const Self = @This();

    input_count: usize,
    hidden_count: usize,
    output_count: usize,
    weights_hidden: []f64,
    biases_hidden: []f64,
    weights_output: []f64,
    biases_output: []f64,

    pub fn predict(
        self: *Self,
        allocator: std.mem.Allocator,
        input: []f64,
    ) ![]f64 {
        const hidden = try allocator.alloc(f64, self.hidden_count);
        defer allocator.free(hidden);

        const output = try allocator.alloc(f64, self.output_count);

        return self.predictInplace(input, hidden, output);
    }

    pub fn predictInplace(
        self: *Self,
        input: []f64,
        hidden: []f64,
        output: []f64,
    ) ![]f64 {
        if (input.len != self.input_count) {
            return RuntimeError.InputSizeMismatch;
        }

        if (hidden.len != self.hidden_count) {
            return RuntimeError.InputSizeMismatch;
        }

        if (output.len != self.output_count) {
            return RuntimeError.InputSizeMismatch;
        }

        for (0..hidden.len) |h| {
            var sum: f64 = 0;
            for (0..input.len) |i| {
                sum += input[i] * self.weights_hidden[i * hidden.len + h];
            }

            hidden[h] = sigmoid(sum + self.biases_hidden[h]);
        }

        for (0..output.len) |o| {
            var sum: f64 = 0;
            for (0..hidden.len) |h| {
                sum += hidden[h] * self.weights_output[h * output.len + o];
            }

            output[o] = sigmoid(sum + self.biases_output[o]);
        }

        return output;
    }

    pub fn print(self: *Self) void {
        std.debug.print("weightsHidden:\n", .{});
        for (0..self.weights_hidden.len) |i| {
            std.debug.print("{d:.6} ", .{self.weights_hidden[i]});
        }
        std.debug.print("\n", .{});

        std.debug.print("biasesHidden:\n", .{});
        for (0..self.biases_hidden.len) |i| {
            std.debug.print("{d:.6} ", .{self.biases_hidden[i]});
        }
        std.debug.print("\n", .{});

        std.debug.print("weightsOutput:\n", .{});
        for (0..self.weights_output.len) |i| {
            std.debug.print("{d:.6} ", .{self.weights_output[i]});
        }
        std.debug.print("\n", .{});

        std.debug.print("biasesOutput:\n", .{});
        for (0..self.biases_output.len) |i| {
            std.debug.print("{d:.6} ", .{self.biases_output[i]});
        }
        std.debug.print("\n", .{});
    }
};

pub const Trainer = struct {
    const Self = @This();

    network: Network,
    hidden: []f64,
    output: []f64,
    grad_hidden: []f64,
    grad_output: []f64,

    pub fn init(
        allocator: std.mem.Allocator,
        input_count: usize,
        hidden_count: usize,
        output_count: usize,
        rand: fn () f64,
    ) !Self {
        const network = Network{
            .input_count = input_count,
            .hidden_count = hidden_count,
            .output_count = output_count,

            .weights_hidden = try allocator.alloc(f64, input_count * hidden_count),
            .biases_hidden = try allocator.alloc(f64, hidden_count),

            .weights_output = try allocator.alloc(f64, hidden_count * output_count),
            .biases_output = try allocator.alloc(f64, output_count),
        };

        for (0..network.weights_hidden.len) |i| {
            network.weights_hidden[i] = rand() - 0.5;
        }
        for (0..network.weights_output.len) |i| {
            network.weights_output[i] = rand() - 0.5;
        }
        for (0..network.biases_hidden.len) |i| {
            network.biases_hidden[i] = 0;
        }
        for (0..network.biases_output.len) |i| {
            network.biases_output[i] = 0;
        }

        const self = Self{
            .network = network,
            .hidden = try allocator.alloc(f64, hidden_count),
            .output = try allocator.alloc(f64, output_count),
            .grad_hidden = try allocator.alloc(f64, hidden_count),
            .grad_output = try allocator.alloc(f64, output_count),
        };

        for (0..self.hidden.len) |i| {
            self.hidden[i] = 0;
        }
        for (0..self.output.len) |i| {
            self.output[i] = 0;
        }
        for (0..self.grad_hidden.len) |i| {
            self.grad_hidden[i] = 0;
        }
        for (0..self.grad_output.len) |i| {
            self.grad_output[i] = 0;
        }

        return self;
    }

    pub fn deinit(self: *Self, allocator: std.mem.Allocator) void {
        allocator.free(self.network.biases_hidden);
        allocator.free(self.network.biases_output);
        allocator.free(self.network.weights_hidden);
        allocator.free(self.network.weights_output);

        allocator.free(self.hidden);
        allocator.free(self.output);
        allocator.free(self.grad_hidden);
        allocator.free(self.grad_output);
    }

    pub fn train(self: *Self, input: []f64, y: []f64, lr: f64) !void {
        if (y.len != self.output.len) {
            return RuntimeError.OutputSizeMismatch;
        }

        _ = try self.network.predictInplace(
            input,
            self.hidden,
            self.output,
        );

        for (0..self.grad_output.len) |o| {
            self.grad_output[o] = (self.output[o] - y[o]) * sigmoidPrime(self.output[o]);
        }

        // ∂L/∂z0 = [(a1 - y) * a1 * (1 - a1)] * w1 * a0 * (1 - a0)
        for (0..self.grad_hidden.len) |h| {
            var sum: f64 = 0;
            for (0..self.grad_output.len) |o| {
                sum += self.grad_output[o] * self.network.weights_output[h * self.output.len + o];
            }

            self.grad_hidden[h] = sum * sigmoidPrime(self.hidden[h]);
        }

        for (0..self.hidden.len) |h| {
            for (0..self.output.len) |o| {
                self.network.weights_output[h * self.output.len + o] -= lr * self.grad_output[o] * self.hidden[h];
            }
        }

        for (0..input.len) |i| {
            for (0..self.hidden.len) |h| {
                self.network.weights_hidden[i * self.hidden.len + h] -= lr * self.grad_hidden[h] * input[i];
            }
        }

        for (0..self.network.biases_output.len) |o| {
            self.network.biases_output[o] -= lr * self.grad_output[o];
        }

        for (0..self.network.biases_hidden.len) |h| {
            self.network.biases_hidden[h] -= lr * self.grad_hidden[h];
        }
    }
};
